ffi/memorymanager.cpp

```diff
@@ -91,26 +99,63 @@ uint8_t *LlvmliteMemoryManager::allocateSection(
             // Remember how much free space is now left in this block
             FreeMB.Free = sys::MemoryBlock((void *)(Addr + Size),
                                            EndOfBlock - Addr - Size);
+            LLVM_DEBUG(dbgs() << format_hex(Addr, 18, true) << "\n");
             return (uint8_t *)Addr;
         }
     }
 
-    // No pre-allocated free block was large enough. Allocate a new memory
-    // region. Note that all sections get allocated as read-write.  The
-    // permissions will be updated later based on memory group.
-    //
+    assert(false && "All memory must be pre-allocated");
+}
+
+static uintptr_t requiredPageSize(uintptr_t Size, uint32_t Alignment) {
+    static const size_t PageSize = sys::Process::getPageSizeEstimate();
+    // Use the same calculation as allocateSection because we need to be able to
+    // satisfy it.
+    uintptr_t RequiredSize =
+        Alignment * ((Size + Alignment - 1) / Alignment + 1);
+    // Round up to the nearest page size. Blocks must be page-aligned.
+    return PageSize * ((RequiredSize + PageSize - 1) / PageSize);
+}
+
+void LlvmliteMemoryManager::reserveAllocationSpace(
+    uintptr_t CodeSize, uint32_t CodeAlign, uintptr_t RODataSize,
+    uint32_t RODataAlign, uintptr_t RWDataSize, uint32_t RWDataAlign) {
+    MemoryGroup &MemGroup = CodeMem;
+
+    // MemoryManager does not have functions for releasing memory after it's
+    // allocated. Normally it tries to use any excess blocks that were
+    // allocated due to page alignment, but if we have insufficient free memory
+    // for the request this can lead to allocating disparate memory that can
+    // violate the ARM ABI. Clear free memory so only the new allocations are
+    // used, but do not release allocated memory as it may still be in-use.
```

Comment:
I had to read this a couple of times, but think I am getting it now. Let me check my understanding/logic here, perhaps it can be used to make the problem description /solution a bit more crisp.

The objective is allocate memory (blocks) that are "near" to each other. Keeping blocks near makes it less likely that the distance between different memory addresses would become too large and e.g. violate ARM ABI relocation restrictions. If a code/rodata/rwdata memory space has been allocated, but not all space is used (e.g. excess blocks that were allocated due to page alignment), then we do mark all memory as being used by clearing the "free memory" here in that space. This has the effect that a next allocation request is not going to try and scavenge some free blocks from somewhere else, thus avoiding that it finds some memory that is potentially "far away".

Reply:
And further to my previous comment, freeing the free space is the crux of the solution. So I think we need to spend a little bit more time on discussing/documenting the alternatives and pros/cons. The alternative mentioned in the [LLVM discourse thread](https://discourse.llvm.org/t/llvm-rtdyld-aarch64-abi-relocation-restrictions/74616/11?u=sjoerdmeijer) talks about potentially doing this in the `finalizeMemory()` method, but doing it here has the benefit of being less intrusive, at the cost of wasting some memory.

It wouldn't be too difficult to quantify the waste, I guess. In an experiment we could iterate over the free blocks and sum the sizes and print that for the numba test suite. Don't know if we are going to learn anything, but just an idea. 

But given the simplicity of the approach, this definitely looks like the most appealing. I am going to look a bit further in this though, to see what you mean by "pending prefix indices" that you mentioned in the description and how that fits into the picture here.